File: pi-extension/src/transport/peer_channel_signed.test.ts
import { EventEmitter } from "node:events";
import { describe, expect, test, vi } from "vitest";
import { generateEd25519Keypair } from "../pairing/crypto.js";
import { signInnerV1 } from "../protocol/signed_inner.js";
import { PlainPeerChannel } from "./peer_channel.js";

class MockRelay extends EventEmitter {
  send = vi.fn();
}

function outerLine(peer: string, inner: unknown): string {
  return JSON.stringify({
    peer,
    ct: Buffer.from(JSON.stringify(inner)).toString("base64"),
  });
}

describe("PlainPeerChannel signed_inner_v1", () => {
  test("wraps outbound ServerMessage when negotiated", () => {
    const relay = new MockRelay();
    const pi = generateEd25519Keypair();
    const app = generateEd25519Keypair();
    const appPk = Buffer.from(app.publicKey).toString("base64");
    const channel = new PlainPeerChannel(
      relay as never,
      appPk,
      "room-1",
      () => undefined,
      undefined,
      {
        localKeypair: pi,
        expectedRemotePubkey: appPk,
        roomId: "room-1",
        requireSigned: true,
      },
    );

    channel.send({ type: "pong", in_reply_to: "p1" });

    const raw = relay.send.mock.calls[0]![0] as string;
    const outer = JSON.parse(raw) as { peer: string; ct: string };
    expect(outer.peer).toBe(appPk);
    const inner = JSON.parse(Buffer.from(outer.ct, "base64").toString("utf8")) as Record<string, unknown>;
    expect(inner["type"]).toBe("signed_inner_v1");
    expect(inner["sender_pk"]).toBe(Buffer.from(pi.publicKey).toString("base64"));
    expect(inner["recipient_pk"]).toBe(appPk);
    expect(inner["room_id"]).toBe("room-1");
    expect(inner["payload_type"]).toBe("pong");
    expect(inner["payload"]).toEqual({ type: "pong", in_reply_to: "p1" });
  });

  test("drops malformed relay frames without throwing", () => {
    const relay = new MockRelay();
    const app = generateEd25519Keypair();
    const appPk = Buffer.from(app.publicKey).toString("base64");
    const messages: unknown[] = [];
    new PlainPeerChannel(
      relay as never,
      appPk,
      "room-1",
      (msg) => messages.push(msg),
    );

    expect(() => relay.emit("message", "not-json")).not.toThrow();
    expect(() => relay.emit("message", JSON.stringify({ peer: appPk }))).not.toThrow();
    expect(() => relay.emit("message", JSON.stringify({ peer: appPk, ct: Buffer.from("not-json").toString("base64") }))).not.toThrow();
    expect(() => relay.emit("message", outerLine(appPk, { id: "no-type" }))).not.toThrow();
    expect(messages).toEqual([]);
  });

  test("unwraps signed inbound and drops unsigned/replayed frames in strict mode", () => {
    const relay = new MockRelay();
    const pi = generateEd25519Keypair();
    const app = generateEd25519Keypair();
    const appPk = Buffer.from(app.publicKey).toString("base64");
    const piPk = Buffer.from(pi.publicKey).toString("base64");
    const messages: unknown[] = [];
    new PlainPeerChannel(
      relay as never,
      appPk,
      "room-1",
      (msg) => messages.push(msg),
      undefined,
      {
        localKeypair: pi,
        expectedRemotePubkey: appPk,
        roomId: "room-1",
        requireSigned: true,
      },
    );

    relay.emit("message", outerLine(appPk, { type: "ping", id: "unsigned" }));
    expect(messages).toEqual([]);

    relay.emit("message", outerLine(appPk, { type: "signed_inner_v1", payload: null }));
    expect(messages).toEqual([]);

    const signed = signInnerV1({
      payload: { type: "ping", id: "p1" },
      sender: app,
      recipientPk: piPk,
      roomId: "room-1",
      now: Date.now(),
      msgId: "msg-1",
    });
    relay.emit("message", outerLine(appPk, signed));
    expect(messages).toEqual([{ type: "ping", id: "p1" }]);

    relay.emit("message", outerLine(appPk, signed));
    expect(messages).toHaveLength(1);
  });
});

File: app/test/data/transport/peer_channel_signed_test.dart
import 'dart:async';
import 'dart:convert';
import 'dart:typed_data';

import 'package:app/data/transport/epk_encoding.dart';
import 'package:app/data/transport/peer_channel.dart';
import 'package:app/pairing/pair_request_flow.dart';
import 'package:app/protocol/protocol.dart';
import 'package:app/protocol/signed_inner.dart';
import 'package:cryptography/cryptography.dart';
import 'package:flutter_test/flutter_test.dart';

class _Q {
  final items = <Uint8List>[];
  final waiters = <Completer<Uint8List>>[];

  void add(Uint8List data) {
    if (waiters.isNotEmpty) {
      waiters.removeAt(0).complete(data);
    } else {
      items.add(data);
    }
  }

  Future<Uint8List> next() {
    if (items.isNotEmpty) return Future.value(items.removeAt(0));
    final c = Completer<Uint8List>();
    waiters.add(c);
    return c.future;
  }
}

class _T implements PeerTransport {
  final _Q sent;
  final _Q recv;
  String activeRoom = 'main';
  _T({required this.sent, required this.recv});

  void setActiveRoom(String roomId) {
    activeRoom = roomId;
  }

  @override
  Future<void> send(Uint8List data) async => sent.add(data);

  @override
  Future<Uint8List> receive() => recv.next();

  @override
  Future<void> close() async {}
}

Future<void> main() async {
  group('PlainPeerChannel signed_inner_v1', () {
    test('propagates initial room into transport', () async {
      final appKey = await Ed25519().newKeyPairFromSeed(Uint8List(32)..[0] = 9);
      final piKey = await Ed25519().newKeyPairFromSeed(Uint8List(32)..[0] = 10);
      final piPk = base64.encode((await piKey.extractPublicKey()).bytes);
      final transport = _T(sent: _Q(), recv: _Q());

      final channel = PlainPeerChannel(
        transport: transport,
        signingKey: appKey,
        expectedRemotePubkey: piPk,
        roomId: 'room-ctor',
        requireSigned: true,
      );

      expect(transport.activeRoom, 'room-ctor');
      await channel.close();
    });

    test('wraps outbound ClientMessage when negotiated', () async {
      final appKey = await Ed25519().newKeyPairFromSeed(Uint8List(32)..[0] = 1);
      final piKey = await Ed25519().newKeyPairFromSeed(Uint8List(32)..[0] = 2);
      final piPk = base64.encode((await piKey.extractPublicKey()).bytes);
      final sent = _Q();
      final recv = _Q();
      final channel = PlainPeerChannel(
        transport: _T(sent: sent, recv: recv),
        signingKey: appKey,
        expectedRemotePubkey: toAppEpk(piPk),
        roomId: 'room-1',
        requireSigned: true,
      );

      await channel.send(Ping(id: 'p1'));

      final frame =
          jsonDecode(utf8.decode(await sent.next())) as Map<String, dynamic>;
      expect(frame['type'], 'signed_inner_v1');
      expect(frame['recipient_pk'], piPk);
      expect(frame['room_id'], 'room-1');
      expect(frame['payload_type'], 'ping');
      expect(frame['payload'], {'type': 'ping', 'id': 'p1'});
    });

    test(
      'unwraps signed inbound and drops unsigned/replayed frames in strict mode',
      () async {
        final appKey = await Ed25519().newKeyPairFromSeed(
          Uint8List(32)..[0] = 3,
        );
        final piKey = await Ed25519().newKeyPairFromSeed(
          Uint8List(32)..[0] = 4,
        );
        final appPk = base64.encode((await appKey.extractPublicKey()).bytes);
        final piPk = base64.encode((await piKey.extractPublicKey()).bytes);
        final sent = _Q();
        final recv = _Q();
        final channel = PlainPeerChannel(
          transport: _T(sent: sent, recv: recv),
          signingKey: appKey,
          expectedRemotePubkey: piPk,
          roomId: 'room-1',
          requireSigned: true,
        );
        final messages = <ServerMessage>[];
        Completer<ServerMessage>? nextMessage;
        final sub = channel.serverMessages.listen((msg) {
          messages.add(msg);
          nextMessage?.complete(msg);
          nextMessage = null;
        });
        Future<ServerMessage> waitNextMessage() {
          final c = Completer<ServerMessage>();
          nextMessage = c;
          return c.future.timeout(const Duration(seconds: 1));
        }

        final signed = await signInnerV1(
          payload: {'type': 'pong', 'in_reply_to': 'p1'},
          senderKey: piKey,
          recipientPk: appPk,
          roomId: 'room-1',
          now: DateTime.now().millisecondsSinceEpoch,
          msgId: 'msg-1',
        );
        final first = waitNextMessage();
        recv.add(
          Uint8List.fromList(
            utf8.encode(
              jsonEncode({'type': 'pong', 'in_reply_to': 'unsigned'}),
            ),
          ),
        );
        recv.add(Uint8List.fromList(utf8.encode(jsonEncode(signed))));
        final firstMsg = await first;
        expect(firstMsg, isA<Pong>());
        expect((firstMsg as Pong).inReplyTo, 'p1');
        expect(messages, hasLength(1), reason: 'unsigned frame is dropped');

        final signed2 = await signInnerV1(
          payload: {'type': 'pong', 'in_reply_to': 'p2'},
          senderKey: piKey,
          recipientPk: appPk,
          roomId: 'room-1',
          now: DateTime.now().millisecondsSinceEpoch,
          msgId: 'msg-2',
        );
        final second = waitNextMessage();
        recv.add(Uint8List.fromList(utf8.encode(jsonEncode(signed))));
        recv.add(Uint8List.fromList(utf8.encode(jsonEncode(signed2))));
        final secondMsg = await second;
        expect(secondMsg, isA<Pong>());
        expect((secondMsg as Pong).inReplyTo, 'p2');
        expect(messages, hasLength(2), reason: 'replay is dropped');

        await sub.cancel();
        await channel.close();
      },
    );
  });
}
